Revert push to v2 official registry

dmcgowan opened the pull request:
No longer push to the official v2 registry when it is available. This allows pulling images from the v2 registry without defaulting push.

Fixes #10557

dmp42 commented:
cc @jfrazelle @crosbymichael @icecrime 
This makes 1.5 behave in the same way as previous versions, like we discussed.

dmcgowan commented:
Updated the logic to match exactly what we had in 1.3 and 1.4 for the official registry. This does not effect private v2 registries.

jfrazelle commented:
cool thanks @dmcgowan

crosbymichael commented:
looks sane so far ;)

icecrime commented:
Just to give a more complete explanation of what's happening there (and hopefully checking in the process that my understanding is correct):

* Push against the official Hub will never use v2: this is the behavior we have with 1.4.1, where push is *always* v1 (be it against public Hub or a private registry).
* Pull against the official Hub of official images only (`repoInfo.Official`) will use v2: this is also the behavior we have in 1.4.1

So in this regard, patch sounds good: waiting on CI and testing locally.

tianon commented:
I've applied this patch locally, but in pulling official images I'm seeing it hit v1.  Isn't that because the new v2 isn't where the official images live, and this v2 code is not using the old python-based v2 API anymore?

jfrazelle commented:
I think they need to make adjustments in production. afaik?

dmcgowan commented:
@tianon: @dmp42 is working on bringing these endpoints online right now.  Until then it will pull from v1

dmp42 commented:
On it.

tianon commented:
```
time="2015-02-03T21:55:32-07:00" level="debug" msg="Calling POST /images/create" 
time="2015-02-03T21:55:32-07:00" level="info" msg="POST /v1.17/images/create?fromImage=debian%3Alatest" 
time="2015-02-03T21:55:32-07:00" level="info" msg="+job pull(debian, latest)" 
time="2015-02-03T21:55:32-07:00" level="info" msg="+job resolve_repository(debian)" 
time="2015-02-03T21:55:32-07:00" level="info" msg="-job resolve_repository(debian) = OK (0)" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="pulling image from host \"docker.io\" with remote name \"library/debian\"" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="pinging registry endpoint https://index.docker.io/v1/" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="attempting v1 ping for registry endpoint https://index.docker.io/v1/" 
time="2015-02-03T21:55:32-07:00" level="info" msg="+job trust_update_base()" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="Fetched 1 base graphs at 2015-02-03 21:55:32.538292526 -0700 MST" 
time="2015-02-03T21:55:32-07:00" level="info" msg="-job trust_update_base() = OK (0)" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="pulling v2 repository with local name \"debian\"" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="pinging registry endpoint https://registry-1.docker.io/v2/" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="attempting v2 ping for registry endpoint https://registry-1.docker.io/v2/" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="hostDir: /etc/docker/certs.d/registry-1.docker.io" 
time="2015-02-03T21:55:32-07:00" level="debug" msg="Reloaded graph with 2 grants expiring at 2015-02-22 22:47:54.722420237 +0000 UTC" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="Getting authorization for library/debian [pull]" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="Pulling tag from V2 registry: \"latest\"" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="[registry] Calling \"GET\" https://registry-1.docker.io/v2/library/debian/manifests/latest" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="https://registry-1.docker.io/v2/library/debian/manifests/latest -- HEADERS: map[User-Agent:[docker/1.4.1-dev go/go1.4.1 git-commit/165ea5c-dirty kernel/3.16.5-gentoo os/linux arch/amd64]]" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="hostDir: /etc/docker/certs.d/registry-1.docker.io" 
time="2015-02-03T21:55:33-07:00" level="info" msg="+job trust_key_check(/library/debian)" 
time="2015-02-03T21:55:33-07:00" level="info" msg="-job trust_key_check(/library/debian) = OK (0)" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="Key check result: \"verified\"" 
time="2015-02-03T21:55:33-07:00" level="info" msg="Image manifest for debian:latest has been verified" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="Image already exists: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="Image already exists: 30d39e59ffe287f29a41a3f8bd70734afc8728329e3289945cbdc5bbf07cd980" 
time="2015-02-03T21:55:33-07:00" level="debug" msg="Image already exists: c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444" 
time="2015-02-03T21:55:33-07:00" level="info" msg="+job log(pull, debian:latest, )" 
time="2015-02-03T21:55:33-07:00" level="info" msg="-job log(pull, debian:latest, ) = OK (0)" 
time="2015-02-03T21:55:33-07:00" level="info" msg="-job pull(debian, latest) = OK (0)" 
```

Looks like v2's getting hit on my machine running with this patch now. :+1:

I would also add that `docker pull -a debian` still appears to be working correctly now that we're hitting v2, so that's also a good sign. :signal_strength:

crosbymichael commented:
Just downloaded an official image and received this:

`debian:jessie: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.`

I guess that is a good sign.

icecrime commented:
Same here, pulling official images look good:

Daemon:
```
DEBU[0010] Pulling tag from V2 registry: "latest"
DEBU[0010] [registry] Calling "GET" https://registry-1.docker.io/v2/library/busybox/manifests/latest
DEBU[0010] https://registry-1.docker.io/v2/library/busybox/manifests/latest -- HEADERS: map[User-Agent:[docker/1.4.1-dev go/go1.4.1 git-commit/22e2254 kernel/3.13.0-24-generic os/linux arch/amd64]]
```

Client:
```
busybox:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.
```

Does it mean that official images have already been rebuilt with the new TarSum version?

icecrime commented:
Pushing looks suspicious:

* It seems to be pushing using v2 endpoints
* I haven't logged in yet, but the message in unrelated

```
DEBU[0301] [registry] Calling "POST" https://registry-1.docker.io/v2/icecrimedev/busybox/blobs/uploads/
DEBU[0301] https://registry-1.docker.io/v2/icecrimedev/busybox/blobs/uploads/ -- HEADERS: map[User-Agent:[docker/1.4.1-dev go/go1.4.1 git-commit/22e2254 kernel/3.13.0-24-generic os/linux arch/amd64]]
DEBU[0301] Using cached token for
DEBU[0301] hostDir: /etc/docker/certs.d/registry-1.docker.io
DEBU[0301] [registry] Calling "PUT"
DEBU[0301]  -- HEADERS: map[User-Agent:[docker/1.4.1-dev go/go1.4.1 git-commit/22e2254 kernel/3.13.0-24-generic os/linux arch/amd64]]
DEBU[0301] Using cached token for
Error pushing to registry: Put ?digest=tarsum.v1%2Bsha256%3Ae3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855: unsupported protocol scheme ""
```

tiborvass commented:
@icecrime Looks like #10557 is the same issue but it's with rc3

icecrime commented:
I guess something server-side must be replying positively to a `/v2/` ping request or we wouldn't be seeing pushes go over v2.

dmcgowan commented:
@icecrime yes,RC3 will attempt a v2 push now that the v2 endpoint is responding but only accepting pulls. You are not seeing that with this PR though right?

icecrime commented:
@dmcgowan Yes my test is with that PR. I think @tiborvass just confirmed that the Hub infra does reply to a /v2/ ping, which makes it eligible for a v2 push as far as I understand.

dmcgowan commented:
Understood that /v2/ accepts a ping which turns on push, but with this PR it should not push to the official registry even with the /v2/ endpoint online.  How are you doing the push, I am not seeing the attempted push with this change.

icecrime commented:
Sorry then, let me double check but I really think I had this PR compiled. However, reading through the code, I don't see where we prevent pushes to the Hub with v2: we create a session (hence doing the ping), and then: 

    if endpoint.Version == registry.APIVersion2 {

dmcgowan commented:
@icecrime the endpoint in this case is actually index.docker.com which is itself only a v1 registry. When we were using the official registry we overwrote the endpoint with the v2 registry endpoint at registry-1.docker.com/v2/. We are going to refactor this code but right now the underlying code has no concept of separate domain endpoints for v1/v2.

icecrime commented:
Ok I'm very sorry I must have got something wrong in my test yesterday, I rebuilt the PR and it's behaving as expected. LGTM.

tiborvass commented on graph/pull.go:
```diff
@@ -72,7 +72,7 @@ func (s *TagStore) CmdPull(job *engine.Job) engine.Status {
 	}
 
-	if len(repoInfo.Index.Mirrors) == 0 && (repoInfo.Index.Official || endpoint.Version == registry.APIVersion2) {
+	if len(repoInfo.Index.Mirrors) == 0 && ((repoInfo.Index.Official && repoInfo.Official) || endpoint.Version == registry.APIVersion2) {
```
@dmcgowan i would put the `repoInfo.Official` first, since that case is rarer.

tiborvass commented:
Apart from small nit, LGTM

dmcgowan committed:
Limit push and pull to v2 official registry

No longer push to the official v2 registry when it is available. This allows pulling images from the v2 registry without defaulting push. Only pull official images from the v2 official registry.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
@tiborvass fixed nit

LK4D4 commented:
LGTM
